README.rst
======================
听云 Python Agent
======================

    该安装包为北京基调网络股份有限公司提供用于应用性能管理的采集客户端，旨在对Python Web应用的全栈性能监测以及采样提供支持。更多资讯请移步听云 `官方网站 <http://www.tingyun.com>`_

安装
------------
使用python包常见安装方法:

    *pip install tingyun*


使用
-------------
    **注**：以下示例，为在linux shell终端下执行。 探针启动时需要的 **配置文件** 环境变量"TING_YUN_CONFIG_FILE"也可以直接设置到操作系统环境里，或者类似示例一样直接设置到当前应用启动的局部环境变量中。

    1、使用听云为用户分发的licenseKey生成配置文件
            *tingyun-admin  generate-config <ACCOUNT-LICENSE-KEY> $config-file-path.ini*

       例如：
            *tingyun-admin  generate-config 111-111-111 /tmp/ty-agent-config.ini*

    2、使用探针启动web应用
            *TING_YUN_CONFIG_FILE=$config-file-path tingyun-admin run-program <APPLICATION-START-COMMAND>*

        例如（启动django应用）：
            *TING_YUN_CONFIG_FILE=/tmp/ty-agent-config.ini tingyun-admin run-program python manage.py runserver 0.0.0.0:8800*

ReleaseNotes.rst
#
## 该文档描述了python 探针版本迭代以及相关功能的变化
## 更多帮助和支持请访问www.tingyun.com 或者直接联系我们的技术支持
#

2018-04-25  Version：1.3.2
    增加：1、增加错误&异常采集
         2、支持自定义事务命名

2017-06-28  Version：1.3.1
    修改：1、优化外部调用逻辑

2017-04-17  Version：1.3.0
    增加：1、支持报表页面查看外部调用应用详情
    增加：2、支持基于pika驱动的RabbitMQ生产者、消费者性能采集，并包含如下功能
            异步、同步模式生产者
            异步、同步模式消费者
            MQ与Web应用之间跨应用

2017-03-17  Version：1.2.0
    增加：1、支持在听云报表情报汇总页查看`应用服务器响应时间分位值`
         2、支持报表设置值，设置`分位值`
    修复：优化Tornado4.x框架(beta 可能会导致应用QPS降低)

2017-2-23  Version：1.1.3
    增加: 增加数据库实例识别
    修改：优化Browser探针自动嵌码逻辑
    修改：增强错误采集准确性
    修复：增强探针稳定性

2016-11-04  Version：1.1.2
    增加：兼容windows python2.6、python2.7环境
    修复：增强探针稳定性

2016-09-08  Version：1.1.1
    增加： 支持python-binary-memcached、pymemcache模块
    修复： 检测端口号时，可能引发问题的潜在bug
    修复： 报表端将探针禁用后，再次启用探针不能启动的问题

2016-08-01  Version：1.1.0
    增加： 支持piston模块
    增加： 支持tornado4，并支持wsgi模式
    修复： flask框架中，因错误参数，导致注册错误处理器失败的问题。


2016-06-24  Version：1.0.6
    增加：外部追踪错误的采集，支持requests、urllib3、urllib2、urllib、httplib2、thrift
    增加：自定义web应用过程命名
    增加：外部请求错误追踪功能
    修复：应用容器父进程中，调用应用导致探针不能捕获性能数据的问题


2016-06-16  Version：1.0.1
    增加：支持代理功能上报数据.
    增加：支持bottle框架，支持范围0.10.x-0.12.x

2016-3-16  Version：0.9.2
    修复: requests包，2.4.1版本以下的兼容性问题
    修复: 错误trace中去除tingyun包信息

2016-1-28  Version：0.9.1
    增加：对以下插件的对外调用，支持听云探针之间的跨应用追踪
        requests
        httplib2
        urllib3
    修复：urllib3中，不能采集get参数的问题
    修复：使用urllib3模块，导致参数解析错误的问题

2015-12-30  Version：0.8.0
    增加：tornado的支持,支持范围v3.x.x-v4.x.x（该框架不支持browser探针自动嵌码）
    增加：verify_certification配置项，用于解决ssl传输数据时校验服务器证书认证失败的问题。
    增加：对flask框架的蓝图组件的性能追踪。

2015-12-02  Version：0.7.7
    增加：支持browser探针自动嵌码
    增加：对browser探针增加配置项

2015-11-20  Version：0.7.5
    增加：gevent部署方式的入口检测，支持范围0.13.0-1.0.2
    增加：应用阻塞时间采集
    增加：多个appName的支持
    增加：在文档中增加对部署探针常见问题的汇总
    修复：错误配置项导致探针无法启动的问题

2015-10-28  Version：0.7.0
    增加：web2py框架的支持，支持范围2.8.1-2.12.x
    增加：开放探针部分API
    增加：增加各个框架版本信息的上报
    修复：flask框架下模板检测的错误

2015-10-27  Version：0.6.7
    修复：用户在开启追踪应用阻塞时间时，数据计算错误的问题
    修复：配置文件因使用中文应用名称导致数据上报失败的问题

2015-10-26  Version：0.6.6
    增加：webpy框架的支持，支持范围0.3.x
    增加：模板mako支持，支持范围v0.7.x-v1.0.x
    增加：一键嵌入探针时若无法找到配置文件，提示未生效该次操作

2015-09-25  Version：0.6.3
    增加：框架flask的支持, 支持范围0.6-0.10
    增加：jinja2模板性能检测的支持, 支持范围2.3以上版本
    修复：默认statuscode造成的服务端数据入库问题
    增加：支持用户私有化探针

2015-08-28  Version：0.5.4
    增加：数据库追踪类型
    增加：cpu、内存的实时采样
    增加：服务器端配置（参数命名应用过程，暂支持url参数）
    增加：外部调用redis、requests、httplib2的支持
    增加：对django框架内部异常采集。

2015-08-06  Version：0.4.3
    修复：在线安装默认配置文件丢失问题
    修复：线程剖析潜在可能崩溃的bug

2015-08-06  Version：0.4.2
    增加：性能剖析功能
    增加：generate-config命令
    增加：源码安装支持跨目录安装
    增加：支持pip、easy_install方式安装
    修复：更详细的python环境信息
    修复：探针内核修复，增加稳定性
    修复：其他小细节以及bug

2015-06-29  Version：0.3.0
    增加：redis的性能数据监测
    增加：mongodb的性能数据监测
    增加：增加check-config命令，用于对配置文件进行校验。


tingyun/embattle/repertoire.py
"""
"""

from tingyun.config.settings import global_settings

tornado_version = ""

try:
    import tornado
    tornado_version = tornado.version
except Exception:
    pass


def defined_repertoire():
    """
    :return:
    """
    hookers = {
        "memcached": [
            {"target": "memcache", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_memcached'},
        ],

        "pymemcache": [
            {"target": "pymemcache.client", 'hook_func': "detect_base_client",
             'hook_module': 'tingyun.armoury.database_pymemcache'},
            {"target": "pymemcache.client", 'hook_func': "detect_pooled_client",
             'hook_module': 'tingyun.armoury.database_pymemcache'},
            {"target": "pymemcache.client.hash", 'hook_func': "detect_has_client",
             'hook_module': 'tingyun.armoury.database_pymemcache'},
        ],
        "bmemcached": [
            {"target": "bmemcached.client", 'hook_func': "detect_client",
             'hook_module': 'tingyun.armoury.database_bmemcache'},
        ],

        # mysql db
        "mysql": [
            {"target": "MySQLdb", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],
        "pymysql": [
            {"target": "pymysql", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],
        "oursql": [
            {"target": "oursql", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],

        # oracle
        "oracle": [
            {"target": "cx_Oracle", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],

        # postgres SQL
        "postgresql": [
            {"target": "postgresql.interface.proboscis.dbapi2", 'hook_func': 'detect',
             'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],
        # postgres SQL
        "psycopg2": [
            {"target": "psycopg2", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],
        "psycopg2ct": [
            {"target": "psycopg2ct", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],
        "psycopg2cffi": [
            {"target": "psycopg2cffi", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],

        # ODBC A Python DB API 2 module for ODBC
        "pyodbc": [
            {"target": "pyodbc", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.database_dbapi2'},
        ],

        # nosql mongodb
        "mongodb": [
            {"target": "pymongo.mongo_client", 'hook_func': 'detect_mongo_client',
             'hook_module': 'tingyun.armoury.database_mongo'},
            {"target": "pymongo.connection", 'hook_func': 'detect_connection',
             'hook_module': 'tingyun.armoury.database_mongo'},
            {"target": "pymongo.collection", 'hook_func': 'detect_collection',
             'hook_module': 'tingyun.armoury.database_mongo'},
        ],

        # nosql redis
        "redis": [
            {"target": "redis.client", 'hook_func': 'detect_client_operation',
             'hook_module': 'tingyun.armoury.database_redis'},
        ],

        # external call
        "urllib": [
            {"target": "urllib", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.external_urllib'},
        ],
        "urllib2": [
            {"target": "urllib2", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.external_urllib2'},
        ],
        # v1.6-v1.14
        "urllib3": [
            {"target": "urllib3.poolmanager", 'hook_func': 'detect', 'hook_module': 'tingyun.armoury.external_urllib3'},
        ],
        # 0.8.0-0.9.3
        "thrift": [
            {"target": "thrift.transport.TSocket", 'hook_func': 'detect_tsocket',
             'hook_module': 'tingyun.armoury.external_thrift'},
            {"target": "thrift.transport.TSSLSocket", 'hook_func': 'detect_tsslsocket',
             'hook_module': 'tingyun.armoury.external_thrift'},
        ],

        # 2.0.0-2.10.0
        "requests": [
            {"target": "requests.sessions", 'hook_func': 'detect_requests_sessions',
             'hook_module': 'tingyun.armoury.external_requests'},
        ],
        # 0.7.5-0.9.2
        "httplib2": [
            {"target": "httplib2", 'hook_func': 'detect_httplib2_http',
             'hook_module': 'tingyun.armoury.external_httplib2'},
            {"target": "httplib2", 'hook_func': 'detect_http_connect_with_timeout',
             'hook_module': 'tingyun.armoury.external_httplib2'},
            {"target": "httplib2", 'hook_func': 'detect_https_connect_with_timeout',
             'hook_module': 'tingyun.armoury.external_httplib2'},
        ],

        # django, this weapon must not be removed.
        "django": [
            {"target": "django.core.handlers.base", 'hook_func': 'detect_middleware',
             'hook_module': 'tingyun.armoury.framework_django'},
            {"target": "django.core.handlers.wsgi", 'hook_func': 'detect_wsgi_entrance',
             'hook_module': 'tingyun.armoury.framework_django'},
            {"target": "django.core.urlresolvers", 'hook_func': 'detect_urlresolvers',
             'hook_module': 'tingyun.armoury.framework_django'},
            {"target": "django.views.generic.base", 'hook_func': 'detect_views_dispatch',
             'hook_module': 'tingyun.armoury.framework_django'},

            {"target": "django.template.loader_tags", 'hook_func': 'detect_template_block_render',
             'hook_module': 'tingyun.armoury.framework_django'},
            {"target": "django.template.base", 'hook_func': 'detect_django_template',
             'hook_module': 'tingyun.armoury.framework_django'},

            {"target": "django.http.multipartparser", 'hook_func': 'detect_http_multipartparser',
             'hook_module': 'tingyun.armoury.framework_django'},

            {"target": "django.core.mail", 'hook_func': 'detect_core_mail',
             'hook_module': 'tingyun.armoury.framework_django'},
            {"target": "django.core.mail.message", 'hook_func': 'detect_core_mail_message',
             'hook_module': 'tingyun.armoury.framework_django'},
        ],

        # django-piston, 0.2.2.1-0.2.3
        "piston": [
            {"target": "piston.resource", 'hook_func': 'detect_piston_resource',
             'hook_module': 'tingyun.armoury.ammunition.piston'},
        ],

        # flask, 0.6-1.0
        "flask": [
            {"target": "flask.app", 'hook_func': 'detect_wsgi_entrance',
             'hook_module': 'tingyun.armoury.framework_flask'},
            {"target": "flask.app", 'hook_func': 'detect_app_entrance',
             'hook_module': 'tingyun.armoury.framework_flask'},
            {"target": "flask.blueprints", 'hook_func': 'detect_app_blueprint_entrance',
             'hook_module': 'tingyun.armoury.framework_flask'},
            {"target": "flask.templating", 'hook_func': 'detect_templates',
             'hook_module': 'tingyun.armoury.framework_flask'},
        ],

        # jinja2 2.3-2.8
        'jinja2': [
            {"target": "jinja2.loaders", 'hook_func': 'detect_template_loader',
             'hook_module': 'tingyun.armoury.template_jinja2'},
            {"target": "jinja2.environment", 'hook_func': 'detect_jinja2',
             'hook_module': 'tingyun.armoury.template_jinja2'},
            ],

        # version 2.8.1-2.12.x
        'web2py': [
            {"target": "gluon.main", 'hook_func': 'detect_wsgi_entrance',
             'hook_module': 'tingyun.armoury.framework_web2py'},
            {"target": "gluon.compileapp", 'hook_func': 'detect_compileapp',
             'hook_module': 'tingyun.armoury.framework_web2py'},
            {"target": "gluon.template", 'hook_func': 'detect_template',
             'hook_module': 'tingyun.armoury.framework_web2py'},
        ],

        # version 0.3.x
        'webpy': [
            {"target": "web.application", 'hook_func': 'detect_wsgi_entrance',
             'hook_module': 'tingyun.armoury.framework_webpy'},

            {"target": "web.application", 'hook_func': 'detect_application',
             'hook_module': 'tingyun.armoury.framework_webpy'},
        ],

        # version 4.x.x
        # when tornado only use as wsgi application, do not detect tornado component
        "tornado4": [
            {"target": "tornado.web", 'hook_func': 'detect_web',
             'hook_module': 'tingyun.armoury.ammunition.tornado.web'},

            {"target": "tornado.httputil", 'hook_func': 'detect_tracker_entrance',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.http_util'},
            {"target": "tornado.httpserver", 'hook_func': 'detect_tracker_export',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.http_server'},
            {"target": "tornado.httpclient", 'hook_func': 'detect_http_client',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.http_client'},
            {"target": "tornado.gen", 'hook_func': 'detect_gen',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.gen'},
            {"target": "tornado.ioloop", 'hook_func': 'detect_ioloop',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.ioloop'},
            {"target": "tornado.concurrent", 'hook_func': 'detect_concurrent',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.concurrent'},
            {"target": "concurrent.futures", 'hook_func': 'detect_concurrent',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.concurrent'},
            {"target": "tornado.stack_context", 'hook_func': 'detect_stack_context',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.stack_context'},
        ]if "4." in tornado_version and not global_settings().tornado_wsgi_adapter_mode else [],

        "tornado-wsgi-adapter-mode": [
            {"target": "tornado.web", 'hook_func': 'detect_handlers',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.wsgi_mode.web'},
            {"target": "tornado.httpclient", 'hook_func': 'detect_http_client',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.wsgi_mode.http_client'},
            {"target": "tornado.wsgi", 'hook_func': 'detect_wsgi_entrance',
             'hook_module': 'tingyun.armoury.ammunition.tornado_4.wsgi_mode.wsgi'},
        ]if "4." in tornado_version and global_settings().tornado_wsgi_adapter_mode else [],

        # mako v0.7.x-v1.0.x
        "mako": [
            {"target": "mako.template", 'hook_func': 'detect_template',
             'hook_module': 'tingyun.armoury.template_mako'},
        ],

        # 0.13.0-1.0.2
        "gevent": [
            {"target": "gevent.pywsgi", 'hook_func': 'detect_wsgi',
             'hook_module': 'tingyun.armoury.matrix.t_gevent'},
            {"target": "gevent.wsgi", 'hook_func': 'detect_pywsgi',
             'hook_module': 'tingyun.armoury.matrix.t_gevent'},
        ],

        # openstack-nova
        'openstack-nova': [
            # nova.api.openstack.wsgi
            {"target": "nova.api.openstack.wsgi", 'hook_func': 'resource_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.api_openstack_wsgi'},

            # nova.wsgi
            {"target": "nova.wsgi", 'hook_func': 'middleware_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.wsgi'},
            {"target": "nova.wsgi", 'hook_func': 'router_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.wsgi'},

            # nova.api.auth
            {"target": "nova.api.auth", 'hook_func': 'keystone_context_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.api_auth'},

            # nova.api.ec2
            {"target": "nova.api.ec2", 'hook_func': 'ec2_keystone_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.api_ec2'},
            {"target": "nova.api.ec2", 'hook_func': 'requestify_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.api_ec2'},
            {"target": "nova.api.ec2", 'hook_func': 'authorizer_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.api_ec2'},
            {"target": "nova.api.ec2", 'hook_func': 'executor_wsgi_entrance',
             'hook_module': 'tingyun.armoury.nova.api_ec2'},

            # webob.client
            {"target": "webob.client", 'hook_func': 'send_request_wsgi_entrance',
             'hook_module': 'tingyun.armoury.webob.client'},
            {"target": "webob.response", 'hook_func': 'response_wsgi_entrance',
             'hook_module': 'tingyun.armoury.webob.response'},
        ],
        # bottle 0.10.x-0.12.x
        "bottle": [
            {"target": "bottle", 'hook_func': 'detect_wsgi_entrance',
             'hook_module': 'tingyun.armoury.framework_bottle'},
            {"target": "bottle", 'hook_func': 'detect_templates',
             'hook_module': 'tingyun.armoury.framework_bottle'},
            {"target": "bottle", 'hook_func': 'detect_app_components',
             'hook_module': 'tingyun.armoury.framework_bottle'},
        ],

        # RabbitMQ pika0.10.x-
        "pika": [
            {"target": "pika.adapters.blocking_connection", 'hook_func': 'detect_block_channel',
             'hook_module': 'tingyun.armoury.mq.mq_pika'},
            {"target": "pika.channel", 'hook_func': 'detect_channel',
             'hook_module': 'tingyun.armoury.mq.mq_pika'},
            {"target": "pika.connection", 'hook_func': 'detect_connection',
             'hook_module': 'tingyun.armoury.mq.mq_pika'},
        ]
    }

    return hookers
